### ISSUE TITLE
Lockless support for parallelism?

### ISSUE THREAD

**Issue by manuelreis** [clab/dynet#748]
Hello,

I am the studying the support for parallelism on Dynet and  I've come across this description:

> Internally, DyNet launches a pool of training processes and automatically handles passing data examples to each worker. Each worker process individually processes a datum, computing the results of the forward and backward passes, computes gradients with respect to each parameter, and passes these results back to the parent process via a shared memory variable. **Whenever the parent process, which is also processing data, completes a gradient computation, it averages all of the gradients currently in the shared memory gradient storage and updates all parameters with respect to that average gradient**. In this way running training on n cores is similar to training with a stochastic minibatch size with expected value of approximately n. This method is quite efficient, achieving nearly linear speedups with increasing numbers of cores, **due to its lockless nature**.

 from here https://github.com/clab/dynet/blob/master/doc/source/multiprocessing.rst

Is this still true? 

Then, I found this snippet of code from https://github.com/clab/dynet/blob/v2.0/dynet/mp.h#L252:

          bool do_update = !header.is_dev_set && cid == 0;
          unsigned counter = 0;
          if (!header.is_dev_set) {
            shared_object->counter_mutex.wait();
            counter = ++shared_object->counter;
            if (do_update) { shared_object->counter = 0; }
            shared_object->counter_mutex.post();
          }
          if (do_update && trainer != nullptr) {
            shared_object->update_mutex.wait();
            trainer->update(1.0 / counter); 
            shared_object->update_mutex.post();
          }
          if (batch_counter == header.report_frequency) {
            if (cid == 0) {
              std::cerr << (header.is_dev_set ? "dev" : "train") << " loss: " << batch_loss << std::endl;
            }
            batch_loss = S();
            batch_counter = 0;
          }
        }
        if (header.end_of_epoch && trainer != nullptr) {
          trainer->update_epoch();
        }

        // Let the parent know that we're done and return the loss value
        write_data(workloads[cid].c2p[1], total_loss);

From my understanding, the _update_mutex_ is acquired before the gradients are updated. So in fact, locks are used?

Also, could you please tell me in which part of the code this is performed by the parent process?

_it averages all of the gradients currently in the shared memory gradient storage and updates all parameters with respect to that average gradient_ 

I could found how you pass the losses from the children to the parents, but not the aforementioned averaging of gradients.

Thanks in advance.

**Comment by armatthews** [comment]
Hi Manuel, 
   You're correct that there is a mutex used, so the claim of being lockless isn't 100% accurate. Nevertheless, the forward and backward passes, which account for the overwhelming majority of computation do not use locks internally.

   The misunderstanding here seems to be what the call to update() does. Gradients are computed by the backward() function, and the gradients are stored in the Parameter objects. The update() call simply loops over all parameters and does the SGD (or Adam, etc.) update on each parameter. This means the lock is only acquired once ever N examples, and is only held for a brief amount of time.

  As for which parts of the code are done by the parent, the parent is the process with cid of 0. The parent also loops over examples and computes gradients. The only difference is that when the parent finishes computing the gradients for an example it calls update() and outputs the loss info to stderr.

  The averaging of gradients is done in two steps. First, note that as backward() computes the gradients for each parameter it sums the new gradients with whatever gradients were already stored in the Parameter object. This means that if you call backward() N times, the gradients stored in the Parameter objects are the sum of the gradients of N examples. Then when we call update, we pass in a scaling factor of 1.0 / counter, where counter counts the number of examples seen since the last update. This has the effect of dividing the gradients by counter, giving us a true average.

**Comment by manuelreis** [comment]
Hi,

First of all thank you very much for your clear answer.

I have still one remaining doubt, since this snippet of code:

                    shared_object->update_mutex.wait();
                    trainer->update(1.0 / counter);
                    shared_object->update_mutex.post();

is only executed by the _parent_ (i.e., is only executed by the worker whose _cid == 0 _), and there is only one parent, why is it wrapped with this _update_mutex_? 
Where is this _update_mutex_ also used and by whom?

Best,

**Comment by manuelreis** [comment]
Hi,

can you please take a look into this issue?
